src/ImageSharp/Processing/Processors/Normalization/AdaptiveHistEqualizationProcessor.cs

```diff
@@ -0,0 +1,230 @@
+﻿// Copyright (c) Six Labors and contributors.
+// Licensed under the Apache License, Version 2.0.
+
+using System;
+using System.Numerics;
+using System.Threading.Tasks;
+using SixLabors.ImageSharp.Advanced;
+using SixLabors.ImageSharp.Memory;
+using SixLabors.ImageSharp.PixelFormats;
+using SixLabors.Memory;
+using SixLabors.Primitives;
+
+namespace SixLabors.ImageSharp.Processing.Processors.Normalization
+{
+    /// <summary>
+    /// Applies an adaptive histogram equalization to the image.
+    /// </summary>
+    /// <typeparam name="TPixel">The pixel format.</typeparam>
+    internal class AdaptiveHistEqualizationProcessor<TPixel> : HistogramEqualizationProcessor<TPixel>
+        where TPixel : struct, IPixel<TPixel>
+    {
+        /// <summary>
+        /// Initializes a new instance of the <see cref="AdaptiveHistEqualizationProcessor{TPixel}"/> class.
+        /// </summary>
+        /// <param name="luminanceLevels">The number of different luminance levels. Typical values are 256 for 8-bit grayscale images
+        /// or 65536 for 16-bit grayscale images.</param>
+        /// <param name="clipHistogram">Indicating whether to clip the histogram bins at a specific value.</param>
+        /// <param name="clipLimitPercentage">Histogram clip limit in percent of the total pixels in the grid. Histogram bins which exceed this limit, will be capped at this value.</param>
+        /// <param name="gridSize">The grid size of the adaptive histogram equalization. Minimum value is 4.</param>
+        public AdaptiveHistEqualizationProcessor(int luminanceLevels, bool clipHistogram, float clipLimitPercentage, int gridSize)
+            : base(luminanceLevels, clipHistogram, clipLimitPercentage)
+        {
+            Guard.MustBeGreaterThanOrEqualTo(gridSize, 4, nameof(gridSize));
+
+            this.GridSize = gridSize;
+        }
+
+        /// <summary>
+        /// Gets the size of the grid for the adaptive histogram equalization.
+        /// </summary>
+        public int GridSize { get; }
+
+        /// <inheritdoc/>
+        protected override void OnFrameApply(ImageFrame<TPixel> source, Rectangle sourceRectangle, Configuration configuration)
+        {
+            MemoryAllocator memoryAllocator = configuration.MemoryAllocator;
+            int numberOfPixels = source.Width * source.Height;
+            Span<TPixel> pixels = source.GetPixelSpan();
+
+            int pixelsInGrid = this.GridSize * this.GridSize;
+            int halfGridSize = this.GridSize / 2;
+            using (Buffer2D<TPixel> targetPixels = configuration.MemoryAllocator.Allocate2D<TPixel>(source.Width, source.Height))
+            {
+                ParallelFor.WithConfiguration(
+                    0,
+                    source.Width,
+                    Configuration.Default,
```

Comment:
Y0ou should use the method's `configuration` argument instead.

Reply:
fixed